src/ppl/src/SimulatedAnnealing.cpp

```diff
@@ -0,0 +1,206 @@
+        // accept new solution, update cost and slots
+        pre_cost = cost;
+        if (prev_slot != -1 && new_slot != -1) {
+          slots_[prev_slot].used = false;
+          slots_[new_slot].used = true;
+        }
+      } else {
+        restorePreviousAssignment(prev_slot, new_slot, pin1, pin2);
+      }
+    }
+
+    temperature *= alpha_;
+  }
+}
+
+void SimulatedAnnealing::getAssignment(std::vector<IOPin>& assignment)
+{
+  for (int i = 0; i < pin_assignment_.size(); i++) {
+    IOPin& io_pin = netlist_->getIoPin(i);
+    Slot& slot = slots_[pin_assignment_[i]];
+
+    io_pin.setPos(slot.pos);
+    io_pin.setLayer(slot.layer);
+    io_pin.setPlaced();
+    assignment.push_back(io_pin);
+    slot.used = true;
+  }
+}
+
+void SimulatedAnnealing::randomAssignment()
+{
+  std::mt19937 g;
+  g.seed(seed_);
+  utl::shuffle(slot_indices_.begin(), slot_indices_.end(), g);
+  for (int i = 0; i < pin_assignment_.size(); i++) {
+    int slot_idx = slot_indices_[i];
+    pin_assignment_[i] = slot_idx;
+    slots_[slot_idx].used = true;
+  }
+}
+
+int SimulatedAnnealing::getAssignmentCost()
+{
+  int cost = 0;
+
+  for (int i = 0; i < pin_assignment_.size(); i++) {
+    int slot_idx = pin_assignment_[i];
+    const odb::Point& position = slots_[slot_idx].pos;
+    cost += netlist_->computeIONetHPWL(i, position);
+  }
+
+  return cost;
+}
+
+void SimulatedAnnealing::perturbAssignment(int& prev_slot,
+                                           int& new_slot,
+                                           int& pin1,
+                                           int& pin2)
+{
+  const float move = distribution_(generator_);
+
+  if (move < swap_pins_) {
+    swapPins(pin1, pin2);
+  } else {
+    movePinToFreeSlot(prev_slot, new_slot, pin1);
+  }
+}
+
+void SimulatedAnnealing::swapPins(int& pin1, int& pin2)
+{
+  pin1 = (int) (std::floor(distribution_(generator_) * num_pins_));
+  pin2 = (int) (std::floor(distribution_(generator_) * num_pins_));
+  std::swap(pin_assignment_[pin1], pin_assignment_[pin2]);
+}
+
+void SimulatedAnnealing::movePinToFreeSlot(int& prev_slot,
+                                           int& new_slot,
+                                           int& pin)
+{
+  pin = (int) (std::floor(distribution_(generator_) * num_pins_));
+  prev_slot = pin_assignment_[pin];
+
+  bool free_slot = false;
+  while (!free_slot) {
+    new_slot = (int) (std::floor(distribution_(generator_) * num_slots_));
+    free_slot = slots_[new_slot].isAvailable() && new_slot != prev_slot;
+  }
+
+  pin_assignment_[pin] = new_slot;
+}
+
+void SimulatedAnnealing::restorePreviousAssignment(const int prev_slot,
+                                                   const int new_slot,
+                                                   const int pin1,
+                                                   const int pin2)
+{
+  if (prev_slot != -1 && new_slot != -1) {
+    // restore single pin to previous slot
```

Comment:
This is requires too much knowledge of how the perturbation works.  Once you extend to groups you'll need a move the knows how to undo itself.  Just noting - it can wait until your next step.

Reply:
My idea is to have a list of pins instead of just two. I'll start this part tomorrow and will keep you updated on my progress.